### PR TITLE
Add/data binding using web socket

### PR DESCRIPTION
**Pull request by Dhruvi-Makwana**

### CODE REVIEW

File: Chatify/chat/api.py [hunk 1]
```diff
@@ -1,21 +1,31 @@
...
-            serializer = UserSerializer(data=request.data)
+            data = request.data.dict()
+            data["mobile_number"] = validate_contact_number(data.get("mobile_number"))
+
+            serializer = UserSerializer(data=data)  # add block except validation as e
             serializer.is_valid(raise_exception=True)
             serializer.save()
             return Response(serializer.data, status=status.HTTP_201_CREATED)
+        except ValidationError as err:
+            print(f"hiiiiiiii {str(err)}")
```
Comment (yatishTT) [inline Chatify/chat/api.py]: why print

File: Chatify/chat/api.py [hunk 2]
```diff
@@ -30,8 +40,46 @@ def get(self, request):
...
 
 
-class UserOnlineAPIView(APIView):
+class LoginAPIView(APIView):
     def post(self, request, *args, **kwargs):
-        user = request.user.id
-        user.save(is_online=True if request.data.get("online") else False)
+        try:
+            login_serializer = LoginSerializer(data=request.data)
+            login_serializer.is_valid()
```
Comment (yatishTT) [inline Chatify/chat/api.py]: raie_exception=True

File: Chatify/chat/api.py [hunk 3]
```diff
@@ -30,8 +40,46 @@ def get(self, request):
...
+        except Exception as e:
+            return Response({"error": str(e)}, status=status.HTTP_400_BAD_REQUEST)
+
+
+class VisibilityStatusAPI(APIView):
+    def post(self, request, *args, **kwargs):
+
+        data = request.data
+        user = request.user
+        data.get("status").lower().strip() == "online"
```
Comment (yatishTT) [inline Chatify/chat/api.py]: why 2 same lines are added

File: Chatify/chat/consumers.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    def websocket_connect(self, event):
+        self.send({"type": "websocket.accept"})
+
+    def websocket_receive(self, event):
+        status = json.loads(event.get("text")).get("status")
+        self.send(
+            {
+                "type": "websocket.send",
+                "text": json.dumps(
+                    {"user_id": self.scope["user"].id, "status": status}
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: if user is online we have to show more data

File: Chatify/chat/static/js/chat.js [hunk 5]
```diff
@@ -1,80 +1,87 @@
...
+	var ws = new WebSocket('ws://127.0.0.1:8000/ws/chat/')
+
+	ws.onopen = function() {
+		}
+	ws.onmessage = function(e) {
+
+		let userDetail = JSON.parse(e.data)
+		if (userDetail.status == "offline") {
+			console.log($scope.chatData)
+			for (i = 0; i <= $scope.chatData.length; i++) {
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: dont't use for loop instead use filter method

Comment (yatishTT) [reply Chatify/chat/static/js/chat.js]: create s separate function to find and remove

File: Chatify/chat/static/js/chat.js [hunk 6]
```diff
@@ -1,80 +1,87 @@
...
+				if ($scope.chatData[i].id == userDetail.user_id) {
+					$scope.chatData.pop(userDetail.user_id)
+				}
+			}
+		}
+	}
+
+	function setUserStatus(status) {
+		ws.send(JSON.stringify({
+			'status': status
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: indentation

File: Chatify/chat/static/js/chat.js [hunk 7]
```diff
@@ -1,80 +1,87 @@
...
-                message: message,
-            });
-        }
-    }
-    $scope.setStatus = function(status) {
-        $scope.status = status;
-    }
+	$scope.ajaxGet = function(url, callback = null) {
+		$http.get(url).then(function(response) {
+			if (callback) {
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: multiple times added same comments don't use $http.get

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/static/js/chat.js (hunk 7 above)
Comment (Dhruvi-Makwana) [reply Chatify/chat/static/js/chat.js]: I saw your comment on Ajaxget, sir, but when I wrote this function, the get method wasn't functioning properly, which is why.
 #However, every other comment is resolved, leaving only this one.
ajaxGet('GET',"/chat/api/userdata/", function(response) {
    $scope.chatData = response.UserData;
    console.log($scope.chatData)
    console.log("####")
})

File: Chatify/chat/consumers.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+from channels.consumer import SyncConsumer, AsyncConsumer
+from channels.generic.websocket import AsyncJsonWebsocketConsumer
+import json
+
+
+class ChatConsumer(SyncConsumer):
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: it should be AsyncSyncConsumer

File: Chatify/chat/routing.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+from django.urls import path, re_path
+from chat.consumers import ChatConsumer, MyConsumer
+
+websocket_urlpatterns = [
+    # path("ws/chat/", ChatConsumer.as_asgi()),
```
Comment (yatishTT) [inline Chatify/chat/routing.py]: remove comment

File: Chatify/chat/consumers.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                "type": "chat.message",
+                "text": json.dumps(json_data),
+            },
+        )
+
+    async def chat_message(self, event):
+        status = json.loads(event.get("text")).get("status")
+        await self.send_json(
+            {
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: create a serializer for this data pass user instance and you will get serialized data

File: Chatify/chat/consumers.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+class MyConsumer(AsyncJsonWebsocketConsumer):
+    async def connect(self):
+        await self.channel_layer.group_add("visiblity-group", self.channel_name)
+        await self.accept()
+
+    async def receive_json(self, event):
+        status = event.get("status")
+        json_data = {
+            "message": event,
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: create a serializer

File: Chatify/chat/consumers.py (hunk 10 above)
Comment (yatishTT) [reply Chatify/chat/consumers.py]: GetUserDataSerializer

File: Chatify/chat/serializers.py [hunk 12]
```diff
@@ -35,5 +34,36 @@ def validate_password(self, password):
...
+
+
+class MessageSerializer(serializers.Serializer):
+    id = serializers.IntegerField()
+    user = serializers.CharField()
+    profile = serializers.URLField()
+    message = serializers.CharField()
+
+
+class GetUserDataSerializer(serializers.ModelSerializer):
```
Comment (yatishTT) [inline Chatify/chat/serializers.py]: UserSerializer

File: Chatify/chat/static/js/chat.js [hunk 13]
```diff
@@ -5,55 +5,65 @@ app.config(function($interpolateProvider) {
...
+
+    $scope.remove = function(chat) {
+        var index = $scope.chatData.indexOf(chat);
+        $scope.chatData.splice(index);
+    }
+    ws.onmessage = function(e, ) {
+        console.log("websocket onmessage open")
+        let userDetail = JSON.parse(e.data)
+        if (userDetail.status == "offline") {
+            for (chat of $scope.chatData) {
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: use filter function

Comment (yatishTT) [reply Chatify/chat/static/js/chat.js]: common function

removeOfflineUser

File: Chatify/chat/static/js/chat.js [hunk 14]
```diff
@@ -5,55 +5,65 @@ app.config(function($interpolateProvider) {
...
+                if (chat["id"] === userDetail.id) {
+                    $scope.remove(chat)
+                }
+            }
+
+        } else if (userDetail.status == "online") {
+            $scope.$apply(function() {
+                $scope.chatData.push(userDetail)
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: Check if user not in the array then only push

addOnlineUserToList

File: Chatify/chat/templates/chat/websocket_connection.html [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+<body>
+    <h1>Chat Applcation</h1>
+    <textarea id="chatlog" cols="50" rows="2"></textarea><br>
+    <input type="text" id="chat-message-input" size="50"><br>
+    <input type="button" value="send" id="chat-message-submit">
+</body>
+<script>
+
+    var ws = new WebSocket('ws://127.0.0.1:8000/chat/sc/')
```
Comment (yatishTT) [inline Chatify/chat/templates/chat/websocket_connection.html]: remove this sample coede

File: Chatify/chat/urls.py [hunk 16]
```diff
@@ -1,18 +1,30 @@
...
+    OnlineUserAPI,
+    LogoutView,
+)
 
+app_name = "chat"
 
 urlpatterns_api = [
     path("api/register/", RegistrationApi.as_view(), name="RegistrationApi"),
     path("api/user/", UserListAPI.as_view(), name="UserListAPI"),
+    path("api/visibility-status/", VisibilityStatusAPI.as_view(), name="online"),
```
Comment (yatishTT) [inline Chatify/chat/urls.py]: still i can see name online in 3rd parameter, I am not in favour of spoon feeding

File: Chatify/chat/urls.py [hunk 17]
```diff
@@ -1,18 +1,30 @@
...
+    path("api/login/", LoginAPIView.as_view(), name="login"),
+    path("chat/api/userdata/", OnlineUserAPI.as_view(), name="userdata"),
```
Comment (yatishTT) [inline Chatify/chat/urls.py]: get_online_users, OnlineUsersAPI, get_online_users

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/Chatify/settings/base.py [hunk 18]
```diff
@@ -202,6 +201,8 @@
     "chat",
 ]
 
+# WSGI_APPLICATION = "Chatify.wsgi.application"
```
Comment (yatishTT) [inline Chatify/Chatify/settings/base.py]: remove comment

File: Chatify/chat/api.py [hunk 19]
```diff
@@ -1,21 +1,32 @@
...
 from .models import User
 from .utils import validate_contact_number
+from django.shortcuts import redirect, reverse
+from django.contrib.auth import authenticate
+from django.contrib.auth import login, logout
+from django.http import JsonResponse
+from rest_framework.serializers import ValidationError
+from channels.layers import get_channel_layer
+from asgiref.sync import async_to_sync
+import json
```
Comment (yatishTT) [inline Chatify/chat/api.py]: looks you are not using black

File: Chatify/chat/api.py [hunk 20]
```diff
@@ -30,8 +41,44 @@ def get(self, request):
...
+            user = authenticate(request=request, **login_serializer.validated_data)
+            if user:
+                login(request, user)
+                user.is_online = True
+                user.save()
+                return Response(
+                    {"login": login_serializer.data}, status=status.HTTP_200_OK
+                )
+            return Response(
+                {"error": "enter a valid username or password"},
```
Comment (yatishTT) [inline Chatify/chat/api.py]: Invalid username or password

Comment (yatishTT) [reply Chatify/chat/api.py]: message should come from constants

File: Chatify/chat/api.py [hunk 21]
```diff
@@ -30,8 +41,44 @@ def get(self, request):
...
-        user.save(is_online=True if request.data.get("online") else False)
+        user = request.user
+        user.is_online = request.data.get("status").lower().strip() == "online"
+        user.save()
         return Response(status=status.HTTP_200_OK)
+
+
+class OnlineUserAPI(APIView):
+    def get(self, request):
+        data = User.objects.filter(is_online=True).order_by("-id")
```
Comment (yatishTT) [inline Chatify/chat/api.py]: is_active=True

File: Chatify/chat/consumers.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+                ),
+            }
+        )
+
+    def websocket_disconnect(self, event):
+        pass
+
+
+class MyConsumer(AsyncJsonWebsocketConsumer):
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: why name is MyConsumer > VisibilityStatusConsumer

File: Chatify/chat/serializers.py [hunk 23]
```diff
@@ -35,5 +34,36 @@ def validate_password(self, password):
         confirmation_password = self.initial_data.get("confirm_password")
         if password and confirmation_password:
             if password != confirmation_password:
-                raise serializers.ValidationError("PASSWORD DOESNOT MATCH")
+                raise serializers.ValidationError("Password doesn't match")
```
Comment (yatishTT) [inline Chatify/chat/serializers.py]: message should be from constants

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/consumers.py [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+        await self.accept()
+
+    async def receive_json(self, event):
+        status = event.get("status")
+        json_data = {
+            "message": event,
+            "id": self.scope["user"].id,
+            "full_name": self.scope["user"].get_full_name(),
+            "profile_photo": self.scope["user"].profile_photo.url,
+            "status": status,
```
Comment (yatishTT) [inline Chatify/chat/consumers.py]: if you are sending from serializers below then why sending from here ?

File: Chatify/chat/serializers.py [hunk 25]
```diff
@@ -35,5 +34,35 @@ def validate_password(self, password):
...
+    status = serializers.SerializerMethodField()
+
+    def get_full_name(self, obj):
+        return f"{obj.first_name} {obj.last_name}"
+
+    def get_profile_photo(self, obj):
+        return obj.profile_photo.url
+
+    def get_status(self, obj):
+        return "offline" if obj.is_online else "online"
```
Comment (yatishTT) [inline Chatify/chat/serializers.py]: "online" if obj.is_online else "offline"

File: Chatify/chat/static/js/chat.js [hunk 26]
```diff
@@ -5,55 +5,70 @@ app.config(function($interpolateProvider) {
...
 app.controller('chatCtrl', function($scope, $http) {
+    var ws = new WebSocket('ws://127.0.0.1:8000/ws/chat/')
+
+    ws.onopen = function() {
+        console.log("websocket connection open")
+    }
+
+    $scope.removeOfflineUser = function(chat) {
+         $scope.chatData = $scope.chatData.filter(data => data.id !== chat.id);
+        $scope.$apply()
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: Why you are not removing

File: Chatify/chat/static/js/chat.js (hunk 13 above)
Comment (yatishTT) [reply Chatify/chat/static/js/chat.js]: userDetail.status == "online" ? $scope.addOnlineUserToList(userDetail) :  $scope.removeOfflineUser(userDetail)

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/consumers.py [hunk 27]
```diff
@@ -8,26 +8,25 @@ async def connect(self):
...
             },
         )
 
+    @sync_to_async
+    def updated_instance(self):
+        from .models import User
+
+        instance = User.objects.get(id=self.scope["user"].id)
+        return instance
+
```
Comment (Dhruvi-Makwana) [inline Chatify/chat/consumers.py:24]: Here i can't get updated user is_online status that's why i have to write this function

File: Chatify/chat/static/js/chat.js (hunk 26 above)
Comment (Dhruvi-Makwana) [reply Chatify/chat/static/js/chat.js]: here only those data return which not match with chat.id. Means its' automatic remove that's why write filter function

Comment (Dhruvi-Makwana) [reply Chatify/chat/static/js/chat.js]: and one more thing if i  write single line function it's not working here with angularJs.

Comment (Dhruvi-Makwana) [reply Chatify/chat/static/js/chat.js]: Screenshot from 2023-05-17 13-39-03

File: Chatify/chat/api.py [hunk 28]
```diff
@@ -1,17 +1,15 @@
...
 from .utils import validate_contact_number
 from django.shortcuts import redirect, reverse
 from django.contrib.auth import authenticate
 from django.contrib.auth import login, logout
 from django.http import JsonResponse
 from rest_framework.serializers import ValidationError
-from channels.layers import get_channel_layer
-from asgiref.sync import async_to_sync
-import json
+from .constants import LOGIN_VALIDATION_ERROR_MESSAGE
```
Comment (Dhruvi-Makwana) [inline Chatify/chat/api.py:12]: use proper flake8 and black.
Screenshot from 2023-05-17 14-35-00

File: Chatify/chat/static/js/chat.js [hunk 29]
```diff
@@ -13,29 +13,23 @@ app.controller('chatCtrl', function($scope, $http) {
...
-            for (chat of $scope.chatData) {
-                if (chat["id"] === userDetail.id) {
-                    $scope.removeOfflineUser(chat)
-                }
-            }
-        } else if (userDetail.status == "online") {
+            $scope.removeOfflineUser(userDetail)
+        }
+        else{
             $scope.addOnlineUserToList(userDetail);
```
Comment (Dhruvi-Makwana) [inline Chatify/chat/static/js/chat.js]: sir now online offline perfectyl work with same browser with different tab. and also in different browser without refresh.
Thank you.

File: Chatify/chat/consumers.py [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+            {
+                "type": "chat.message",
+            },
+        )
+
+    @sync_to_async
+    def updated_instance(self):
+        from .models import User
+
+        instance = User.objects.get(id=self.scope["user"].id)
```
Comment (yatishTT) [inline Chatify/chat/consumers.py:22]: try 
user = self.scope["user"]
user.refresh_from_db()

File: Chatify/chat/static/js/chat.js (hunk 26 above)
Comment (yatishTT) [reply Chatify/chat/static/js/chat.js]: don't write if

**Review by yatishTT — CHANGES_REQUESTED** [review]

**Review by yatishTT — APPROVED** [review]

**Pull request merged**